File: frontend/src/lib/api/albaranes.ts
import { request, fetchText } from "./client";

export interface DeliveryNoteLine {
    id: string;
    product_id: string | null;
    description: string;
    quantity: number;
    unit_price: number;
    tax_percentage: number;
    total: number;
}

export interface DeliveryNote {
    id: string;
    tenant_id: string;
    client_id: string | null;
    albaran_number: string;
    date: string;
    status: "draft" | "confirmed" | "recibido" | "en_proceso" | "listo" | "delivered" | "anulado";
    /** T7: facturas enlazadas (vacío = sin facturar). */
    invoice_ids?: string[];
    notes: string | null;
    amount_base: number;
    tax_amount: number;
    amount_total: number;
    created_at: string | null;
    lines: DeliveryNoteLine[];
}

export interface DeliveryNoteCreate {
    client_id?: string;
    client_name?: string;
    date?: string;
    notes?: string;
    lines: { product_id?: string; description: string; quantity: number; unit_price: number; tax_percentage: number }[];
}

export const albaranes = {
    list: () => request<DeliveryNote[]>("/api/v1/albaranes"),
    get: (id: string) => request<DeliveryNote>(`/api/v1/albaranes/${id}`),
    create: (data: DeliveryNoteCreate) => request<DeliveryNote>("/api/v1/albaranes", { method: "POST", body: JSON.stringify(data) }),
    updateStatus: (id: string, status: string) => request<DeliveryNote>(`/api/v1/albaranes/${id}/status`, { method: "PATCH", body: JSON.stringify({ status }) }),
    /** HTML del ticket-resguardo 80 mm (vertical tintorería). */
    ticketHtml: (id: string): Promise<string> => fetchText(`/api/v1/albaranes/${id}/ticket`),
    /** T7: factura agrupada — N albaranes del mismo cliente → 1 factura borrador. */
    facturar: (albaran_ids: string[]) =>
        request<{ id: string; invoice_number: string | null }>("/api/v1/albaranes/facturar", {
            method: "POST",
            body: JSON.stringify({ albaran_ids }),
        }),
    delete: (id: string) => request(`/api/v1/albaranes/${id}`, { method: "DELETE" }),
    pdfUrl: (id: string) => `/api/v1/albaranes/${id}/pdf`,
};

File: backend/app/api/v1/schemas/albaranes.py
"""Pydantic schemas for albaranes (delivery notes)."""

from datetime import date as date_type
from datetime import datetime
from uuid import UUID

from pydantic import BaseModel


class DeliveryNoteLineCreate(BaseModel):
    product_id: UUID | None = None
    description: str
    quantity: float = 1
    unit_price: float = 0
    tax_percentage: float = 21


class DeliveryNoteCreate(BaseModel):
    client_id: UUID | None = None
    client_name: str | None = None
    date: date_type | None = None
    notes: str | None = None
    lines: list[DeliveryNoteLineCreate] = []


class DeliveryNoteUpdate(BaseModel):
    """Edición T8. None = no tocar; "" en notes/client_name = limpiar."""

    client_id: UUID | None = None
    client_name: str | None = None
    date: date_type | None = None
    notes: str | None = None
    lines: list[DeliveryNoteLineCreate] | None = None


class DeliveryNoteStatusUpdate(BaseModel):
    status: str  # draft, confirmed, delivered


class DeliveryNoteLineResponse(BaseModel):
    id: UUID
    product_id: UUID | None = None
    description: str
    quantity: float
    unit_price: float
    tax_percentage: float
    total: float
    model_config = {"from_attributes": True}


class DeliveryNoteResponse(BaseModel):
    id: UUID
    tenant_id: UUID
    client_id: UUID | None = None
    albaran_number: str
    date: date_type
    status: str
    notes: str | None = None
    amount_base: float
    tax_amount: float
    amount_total: float
    created_at: datetime | None = None
    lines: list[DeliveryNoteLineResponse] = []
    # T7: facturas enlazadas (lo puebla list_albaranes; vacío en otras rutas).
    invoice_ids: list[UUID] = []
    # Datos del cliente (solo en el listado): buscador por nombre/NIF/tel/email.
    client_name: str | None = None
    client_nif: str | None = None
    client_phone: str | None = None
    client_email: str | None = None
    model_config = {"from_attributes": True}


class FacturarAlbaranesRequest(BaseModel):
    albaran_ids: list[UUID]
